presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -53,17 +58,31 @@ protected Optional<PlanNodeStatsEstimate> doCalculate(AggregationNode node, Stat
             return Optional.empty();
         }
 
-        if (node.getStep() != SINGLE) {
-            return Optional.empty();
+        boolean isAggregateToSingleRow = isAtMostScalar(node, lookup);
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+
+        // Set estimation boundary for aggregation:
+        // If the source only return a single row, it's guaranteed the upper bound is a single row output for this agg.
+        // Aggregation node cannot return output row count more than estimated row count of source node.
+        // So if source stats is present, use the output row count from the source stats as the estimated output row
+        // count as the upper bound limit to avoid returning an unknown output potentially breaks the join search space
+        if (Double.compare(sourceStats.getOutputRowCount(), SINGLE_ROW_COUNT) == 0) {
```

Comment:
we do the min anyway of source and the computed value, so we shouldn't need to do this.